internal/yggdrasil/yggdrasil.go

```diff
@@ -416,3 +432,129 @@ func (h *Handler) setStorageConfiguration(ctx context.Context,
 
 	return err
 }
+
+func (h *Handler) readAndValidateSecret(ctx context.Context, secretName, secretNamespace string, secretKeys keyMapType) (*corev1.Secret, error) {
+	optional := secretKeys == nil
+	secretObj := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: secretNamespace}}
+	err := h.client.Get(ctx, client.ObjectKeyFromObject(secretObj), secretObj)
+	if err != nil {
+		if errors.IsNotFound(err) && optional {
+			return nil, nil
+		}
+		return nil, err
+	}
+	for key := range secretKeys {
+		if secretObj.Data != nil {
+			if _, ok := secretObj.Data[key]; ok {
+				continue
```

Comment:
This part it's not tested, no?

Reply:
We don't test all use-cases/permutations in yggdrasil.go and in the entire project in general

Reply:
Good catch @ydayagi! We can improve this. It is  good practice to find bugs with unit tests rather than production.

Reply:
Let's improve coverage. Would you like to do it in this PR or new one?

Reply:
this PR

Reply:
this code is already covered. possitive test is "Secrets reading succeeded". Negative test is "Secrets missing secret key". So we verify that if key is missing continue does not happen. we verify that if key is present there's no error.

Reply:
I think new PR would be better to improve test coverage but it is up to you, This PR is OK as well.

Reply:
> This part it's not tested, no?

I was mistaken. It is covered